src/tql/magic.py

```diff
@@ -0,0 +1,121 @@
+from IPython.core.magic_arguments import argument, magic_arguments, parse_argstring
+from typedb.common.exception import TypeDBClientException
+import tql.connection
+import tql.query
+from tql.exception import QueryParsingError
+
+
+def substitute_vars(query, local_ns):
+    try:
+        query_vars = "".join(query.split("\"")[::2]).replace("{", "}").split("}")[1::2]
+    except IndexError:
+        return query
+
+    for var in query_vars:
+        try:
+            val = local_ns[var]
+        except KeyError:
+            raise QueryParsingError("No variable found in local namespace with name: {}".format(var))
+
+        if type(val) is str:
+            val = "\"{}\"".format(val.replace("\"", "'"))
+        else:
+            val = str(val)
+
+        query = query.replace("{" + var + "}", val)
+
+    return query
+
+
+@magics_class
+class TQLMagic(Magics, Configurable):
+    show_connection = Bool(
+        True,
+        config=True,
+        help="Always show current connection name when executing a query."
+    )
+    create_database = Bool(
+        True,
+        config=True,
+        help="Create database when opening a connection if it does not already exist."
+    )
+    strict_transactions = Bool(
+        False,
+        config=True,
+        help="Require session and transaction types to be specified for every transaction."
+    )
+    global_inference = Bool(
+        False,
+        config=True,
+        help="Enable rule inference for all queries. Can be overridden per query with -i."
+    )
+
+    @needs_local_scope
+    @line_magic("tql")
+    @cell_magic("tql")
+    @magic_arguments()
+    @argument("line", default="", nargs="*", type=str, help="Valid TypeQL string.")
+    @argument("-a", "--address", type=str, help="TypeDB server address for new connection.")
+    @argument("-d", "--database", type=str, help="Database name for new connection.")
+    @argument("-n", "--alias", type=str, help="Custom name for new connection, or name of existing connection to select.")
+    @argument("-l", "--connections", action="store_true", help="List currently open connections.")
+    @argument("-c", "--close", type=str, help="Close a connection by name.")
+    @argument("-k", "--delete", type=str, help="Close a connection by name and delete its database.")
+    @argument("-p", "--parallelisation", type=int, help="Number of server communication threads for new connection.")
+    @argument("-r", "--result", type=str, help="Assign query result to the named variable instead of printing.")
+    @argument("-f", "--file", type=str, help="Read in query from a TypeQL file at the specified path.")
+    @argument("-i", "--inference", type=bool, help="Enable (True) or disable (False) rule inference for query.")
+    @argument("-s", "--session", type=str, help="Force a particular session type for query, 'schema' or 'data'.")
+    @argument("-t", "--transaction", type=str, help="Force a particular transaction type for query, 'read' or 'write'.")
```

Comment:
It feels as though we're offering slightly too much through this interface - I think a natural way to split this up would be into TypeDB connection and configuration statements and the actual running of TypeQL statements.

For example, we would open a connection with `%typedb -a localhost:1729 -d iam -p 4` and run a query with `%tql -i True -f schema.tql`. 

The command `%tql -c iam@localhost:1729 match $x sub thing;` closes the connection and does nothing else. This is the 'correct' behaviour, but it feels strange that such a statement is valid. I think we can more tightly scope both commands to reduce possible misuse/misinterpretation by splitting this out as I described above. We'd achieve extensibility in this way as well, as the number of possible arguments for this `%tql` command is already quite substantial.

Reply:
This seems logical to me, but I anticipate potential issues. I'll give it a go and get back to you.